packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -109,248 +111,149 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
-    return (
-      lhs.id === rhs.id &&
-      (lhs.className === rhs.className ||
-        (await classHierarchy.classDerivesFrom(lhs.className, rhs.className)) ||
-        (await classHierarchy.classDerivesFrom(rhs.className, lhs.className)))
+            }
+            return undefined;
+          }),
+          filter((def): def is GenericHierarchyNodeDefinition => !!def),
+        ),
+
+        instanceNodeDefinitions.pipe(
+          mergeMap((definition) => {
+            if (filteringHelper.hasFilterTargetAncestor) {
+              // if we have a filter target ancestor, we don't need to filter the definitions - we use all of them
+              return of(applyECInstanceIdsSelector(definition));
+            }
+            return defer(async () => {
+              const imodelAccess = this._imodelAccess;
+              const matches = new Map<Id64String, IModelInstanceKey[]>();
+              async function getEntries(x: IModelInstanceKey) {
+                let entries = matches.get(x.id);
+                if (!entries) {
+                  entries = [];
+                  matches.set(x.id, entries);
+                }
+                // make sure that we don't have path identifiers that only differ by class name, where
+                // class names are in the same class hierarchy, e.g.:
+                // [{ className: "bis.Element", id: "0x1" }, { className: "bis.Subject", id: "0x1" }]
+                // are actually the same instance, so we only need to keep one of them, or otherwise the query will
+                // duplicate it
+                for (const entry of entries) {
+                  if (
+                    entry.className === x.className ||
+                    (await imodelAccess.classDerivesFrom(entry.className, x.className)) ||
+                    (await imodelAccess.classDerivesFrom(x.className, entry.className))
+                  ) {
+                    return undefined;
+                  }
+                }
+                return entries;
+              }
+              for (const id of childNodeFilteringIdentifiers) {
+                // take only identifiers that match the definition by type, source and class
+                if (!HierarchyNodeIdentifier.isInstanceNodeIdentifier(id)) {
+                  continue;
+                }
+                if (id.imodelKey && id.imodelKey !== imodelAccess.imodelKey) {
+                  continue;
+                }
+                if (id.className !== definition.fullClassName) {
+                  if (
+                    !(await Promise.all([
+                      imodelAccess.classDerivesFrom(id.className, definition.fullClassName),
+                      imodelAccess.classDerivesFrom(definition.fullClassName, id.className),
+                    ]).then(([isDerivedFrom, isDerivedTo]) => isDerivedFrom || isDerivedTo))
+                  ) {
```

Comment:
Do you think it's better to have nested ifs instead of a single if:
```ts
if (
  id.className !== definition.fullClassName &&
  !(await Promise.all([
    imodelAccess.classDerivesFrom(id.className, definition.fullClassName),
    imodelAccess.classDerivesFrom(definition.fullClassName, id.className),
  ]).then(([isDerivedFrom, isDerivedTo]) => isDerivedFrom || isDerivedTo))
) {
  continue;
}
```

Reply:
Both aren't great, but maybe nested conditions are worse... Applied your suggestion.